egs2/TEMPLATE/speechlm1/speechlm.sh

```diff
@@ -0,0 +1,734 @@
+        _opts+="--skip_sum_stats"
+        ${python} -m espnet2.bin.aggregate_stats_dirs ${_opts} --output_dir "${speechlm_stats_dir}"
+
+        # (Jinchuan) we only care about the #frames / #patches.
+        for module in enc dec; do
+            for dset in train valid; do
+                if [ -f ${speechlm_stats_dir}/${dset}/${module}_seq_shape ]; then
+                    cat ${speechlm_stats_dir}/${dset}/${module}_seq_shape |\
+                    awk -F ',' '{print $1}' \
+                    > ${speechlm_stats_dir}/${dset}/${module}_seq_lengths
+                fi
+            done
+        done
+    fi
+
+
+    if [ ${stage} -le 8 ] && [ ${stop_stage} -ge 8 ]; then
+        _train_dir="${data_feats}/${train_set}"
+        _valid_dir="${data_feats}/${valid_set}"
+        log "Stage 8: SpeechlLM training: train_jsons=${train_jsons}, valid_set=${valid_jsons}"
+
+        _opts=
+        if [ -n "${train_config}" ]; then
+            # To generate the config file: e.g.
+            #   % python3 -m espnet2.bin.asr_train --print_config --optim adam
+            _opts+="--config ${train_config} "
+        fi
+
+        _data_opts+="--train_shape_file ${speechlm_stats_dir}/train/dec_seq_lengths "
+        _data_opts+="--valid_shape_file ${speechlm_stats_dir}/valid/dec_seq_lengths "
+        if [ -f ${speechlm_stats_dir}/train/enc_seq_shape ]; then
+            _data_opts+="--train_shape_file ${speechlm_stats_dir}/train/enc_seq_lengths "
+        fi
+        if [ -f ${speechlm_stats_dir}/valid/enc_seq_shape ]; then
+            _data_opts+="--valid_shape_file ${speechlm_stats_dir}/valid/enc_seq_lengths "
+        fi
+        
+        log "Generate '${speechlm_exp}/run.sh'. You can resume the process from stage 7 using this script"
+        mkdir -p "${speechlm_exp}"; echo "${run_args} --stage 7 \"\$@\"; exit \$?" > "${speechlm_exp}/run.sh"; chmod +x "${speechlm_exp}/run.sh"
+
+        log "SpeechLM training started... log: '${speechlm_exp}/train.log'"
+        if echo "${cuda_cmd}" | grep -e queue.pl -e queue-freegpu.pl &> /dev/null; then
+            # SGE can't include "/" in a job name
+            jobname="$(basename ${speechlm_exp})"
+        else
+            jobname="${speechlm_exp}/train.log"
+        fi
+        # shellcheck disable=SC2086
+        ${python} -m espnet2.bin.launch \
+            --cmd "${cuda_cmd} --name ${jobname}" \
+            --log "${speechlm_exp}"/train.log \
+            --ngpu "${ngpu}" \
+            --num_nodes "${num_nodes}" \
+            --init_file_prefix "${speechlm_exp}"/.dist_init_ \
+            --multiprocessing_distributed true -- \
+            ${python} -m "espnet2.bin.speechlm_train" \
+                --use_preprocessor true \
+                --token_list ${token_list_dir}/token_list \
+                --token_bias ${token_list_dir}/token_bias.json \
+                --non_linguistic_symbols "${nlsyms_txt}" \
+                --cleaner "${cleaner}" \
+                --g2p "${g2p}" \
```

Comment:
Not pretty sure, but do you think cleaner and g2p would be a good note for this part? For TTS, it should be no problem, but I feel it might be limited when extending to other tasks in the future.